tests/unit/core/test_git_shell.py

```diff
@@ -42,3 +49,138 @@ def test_check_git_dir(tmp_path):
+
+    # AND a second commit
+    second_file = repo.path / "second.py"
+    second_content = "Second file (not included)"
+    second_file.write_text(second_content)
+    repo.add("second.py")
+    repo.create_commit()
+
+    # WHEN scanning since the second commit
+    filepaths = get_filepaths_from_ref(tmp_path, "HEAD~1")
+
+    # THEN file from first commit is part of filepaths
+    assert "first.py" in filepaths
+    # THEN file from first commit is not part of filepaths
+    assert "second.py" not in filepaths
+
+
+def test_get_staged_filepaths(tmp_path):
+    # GIVEN a repository
+    repo = Repository.create(tmp_path)
+    repo.create_commit()
+
+    # AND a first commit
+    first_file = repo.path / "first.py"
+    first_content = "First file (included)"
+    first_file.write_text(first_content)
+    repo.add("first.py")
+    repo.create_commit()
+
+    # AND staged content
+    second_file = repo.path / "second.py"
+    second_content = "Second file (included)"
+    second_file.write_text(second_content)
+    repo.add("second.py")
+
+    # WHEN scanning for files, including staged
+    filepaths = get_staged_filepaths(tmp_path)
+
+    # THEN file from first commit is part of filepaths
+    assert "first.py" in filepaths
+    # THEN file from first commit is part of filepaths
+    assert "second.py" in filepaths
+
+
+def test_tar_from_ref_and_filepaths(tmp_path):
+    # GIVEN a repository
+    repo = Repository.create(tmp_path)
+    repo.create_commit()
+
+    first_file_name = "first.py"
+    first_ignored_file_name = "first_ignored.py"
+    second_file_name = "second.py"
+
+    # AND a first commit
+    first_file = repo.path / first_file_name
+    first_content = "First file (included)"
+    first_file.write_text(first_content)
+    repo.add(first_file_name)
+
+    first_ignored_file = repo.path / first_ignored_file_name
+    first_ignored_content = "First file (filtered out)"
+    first_ignored_file.write_text(first_ignored_content)
+    repo.add(first_ignored_file_name)
+    repo.create_commit()
+
+    # AND a second commit
+    second_file = repo.path / second_file_name
+    second_content = "Second file (not included)"
+    second_file.write_text(second_content)
+    repo.add(second_file_name)
+    repo.create_commit()
+
+    # AND a filter function
+    def filter(path, content):
+        return "ignored" not in path
+
+    # AND a list of filepaths
+    filepaths = [first_file_name, first_ignored_file_name]
+
+    # WHEN creating a tar
+    tarbytes = tar_from_ref_and_filepaths(tmp_path, "HEAD~1", filepaths, filter)
+
+    tar_stream = BytesIO(tarbytes)
+    with tarfile.open(fileobj=tar_stream, mode="r:gz") as tar:
+        filenames = tar.getnames()
+
+        # THEN first file in in tar
+        assert first_file_name in filenames
+        # THEN first_ignored file in not in tar
+        assert first_ignored_file_name not in filenames
+        # THEN second file in not in tar
+        assert second_file_name not in filenames
```

Comment:
Unless I am mistaken, this can be simplified as `assert filenames == [first_file_name]`, no?